src/types.rs

```diff
@@ -261,3 +262,59 @@ impl BufRingEntry {
         &(*ring_base).0.resv
     }
 }
+
+/// A destination slot for sending fixed resources
+/// (e.g. [`opcode::MsgRingSendFd`](crate::opcode::MsgRingSendFd)).
+#[derive(Debug, Clone, Copy)]
+pub struct DestinationSlot {
+    dest: KernelIndexDest,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum KernelIndexDest {
+    /// Equivalent to `IORING_FILE_INDEX_ALLOC`.
+    Auto,
+    /// Fixed slot as indexed by the kernel (target+1).
+    Index(u32),
+}
+
+impl DestinationSlot {
+    const AUTO_ALLOC: u32 = sys::IORING_FILE_INDEX_ALLOC as u32;
+
+    /// Use an automatically allocated target slot.
+    pub fn auto_target() -> Self {
+        Self {
+            dest: KernelIndexDest::Auto,
+        }
+    }
+
+    /// Try to use a given target slot.
+    ///
+    /// Valid slots are in the range from `0` to `u32::MAX - 2` inclusive.
+    pub fn try_from_slot_target(target: u32) -> io::Result<Self> {
+        const MAX_INDEX: u32 = DestinationSlot::AUTO_ALLOC - 2;
+        match target {
+            v if v <= MAX_INDEX => Ok(Self {
+                dest: KernelIndexDest::Index(v + 1),
+            }),
+            v => Err(io::Error::new(
```

Comment:
return `Result<Self, u32> is enough, I prefer to have `io::Error` be treated as a system error

Reply:
I feel that it gets harder for consumers to decipher, but indeed I can switch to that. What is the `u32` in the `Err` variant? The input argument reflected back? If so, would you prefer a `Result<Self, ()>` maybe?

Reply:
I think it is common to return to input itself when creation fails. returnin `Option` or `Result<_, ()>` is also good. Or define a simple struct `InvalidInput`.